feat: add base of mini ast

kmw14641 opened the pull request:
- from AST of rustc, only extract structure that will need for rust to ts (i.e. mini AST)
- commented todo features
- key challenge for this pr: how to make string source to this ast (related to mini lexer and parser)
- I tried to reduce depth of tree as much as possible, but result was not that much because some structures look meaningful, like plural structure with just vector that catch eyes. welcome for opinions.

kmw14641 committed:
feat: add base of mini ast

ChrisCho-H commented on src/ast.rs:
```diff
@@ -0,0 +1,92 @@
+}
+
+pub struct Fn {
+    pub generics: Generics,
```
Is it ok not to include `defaultness`? `Fn` in rustc ast does.

kmw14641 replied:
I thought that there is no way to change that defaultness when I saw rust syntax docs. Actually, I couldn't understand the PR of function defaultness feature deeply. It looks like making final function which cannot override more, but figuring out how to make it final or what situation needs this in rust may need more time. How about change this when needs occurs while parsing?

ChrisCho-H commented on src/ast.rs:
```diff
@@ -0,0 +1,92 @@
+
+pub struct Generics {
+    pub params: Vec<GenericParam>,
+    pub where_clause: Vec<GenericParam>,
```
I think it would be better to represent `where_clause` in a similar form from rustc ast, which is
```rust
/// A single predicate in a where-clause.
#[derive(Clone, Encodable, Decodable, Debug)]
pub enum WherePredicate {
    /// A type bound (e.g., `for<'c> Foo: Send + Clone + 'c`).
    BoundPredicate(WhereBoundPredicate),
    /// A lifetime predicate (e.g., `'a: 'b + 'c`).
    RegionPredicate(WhereRegionPredicate),
    /// An equality predicate (unsupported).
    EqPredicate(WhereEqPredicate),
}
```

In rustc ast, `WhereClause` just contains everything within `WherePredicate`, except `has_where_token` which doesn't seem to be required for us.

kmw14641 replied:
lifetime doesn't exist in ts, and not need to support unsupported, so I modified it with just one variant.
It make sense to maintain original structure as much as possible because need of more information is possible and also unnecessary preprocessing is possible, and original structure can role as standard of this.
Let's make AST bigger or smaller after some progress of parsing or transpiling.

ChrisCho-H commented on src/ast.rs:
```diff
@@ -0,0 +1,92 @@
+    pub decl: FnDecl,
+}
+
+pub struct FnHeader {
```
Please include safety, constness and ext, with comments as well.
```rust
/// A function header.
///
/// All the information between the visibility and the name of the function is
/// included in this struct (e.g., `async unsafe fn` or `const extern "C" fn`).
#[derive(Clone, Copy, Encodable, Decodable, Debug)]
pub struct FnHeader {
    /// Whether this is `unsafe`, or has a default safety.
    pub safety: Safety,
    /// Whether this is `async`, `gen`, or nothing.
    pub coroutine_kind: Option<CoroutineKind>,
    /// The `const` keyword, if any
    pub constness: Const,
    /// The `extern` keyword and corresponding ABI string, if any.
    pub ext: Extern,
}
```

kmw14641 replied:
As mentioned above, also accepted your opinion.

ChrisCho-H reviewed (changes requested):
It's always good to have **AS MANY COMMENTS AS POSSIBLE**, to log history and make it understandable for whom didn't code this.

kmw14641 committed:
fix: add WhereClause and FnHeader, add Rust's and my comment

kmw14641 commented:
Reflected change request!
Also added comments from original Rust AST (which I thought that unnecessary because already github exists, but rethought that explaining code doesn't need to have dependency to other repository), and added some comments to changes from original AST, but may need to be more detailed.
